src/jit/optimizer.cpp

```diff
@@ -1790,7 +1774,268 @@ void Compiler::optFindNaturalLoops()
                 }
 #endif // FEATURE_EH_FUNCLETS && defined(_TARGET_ARM_)
 
-                /* At this point we have a loop - record it in the loop table
+                // Compact the loop (sweep through it and move out any blocks that aren't part of the
+                // flow cycle), and count the exits.
+                BasicBlock* moveAfter = nullptr;
+                for (BasicBlock* previous = top->bbPrev; previous != bottom;)
+                {
+                    BasicBlock* block = previous->bbNext;
+
+                    if (!loopBlocks.isMember(block->bbNum))
+                    {
+                        // This code is lexically between TOP and BOTTOM, but it does not
+                        // participate in the flow cycle.  Check for a run of consecutive
+                        // such blocks.
+                        BasicBlock* lastExitBlock = block;
+                        BasicBlock* nextLoopBlock = block->bbNext;
+                        while (!loopBlocks.isMember(nextLoopBlock->bbNum))
+                        {
+                            lastExitBlock = nextLoopBlock;
+                            nextLoopBlock = nextLoopBlock->bbNext;
+                            // This loop must terminate because we know BOTTOM is in loopBlocks.
+                        }
+
+                        if (moveAfter == nullptr)
+                        {
+                            // Find an insertion point for blocks we're going to move.  Move them down
+                            // out of the loop, and if possible find a spot that won't break up fall-through.
+                            moveAfter = bottom;
+                            while (moveAfter->bbFallsThrough())
+                            {
+                                // Avoid splitting up this fall-through if we can.
+                                BasicBlock* next = moveAfter->bbNext;
+
+                                if (!BasicBlock::sameEHRegion(moveAfter, next))
+                                {
+                                    // Don't cross an EH region boundary
+                                    break;
+                                }
+
+                                if ((next->bbJumpKind == BBJ_ALWAYS) || (next->bbJumpKind == BBJ_COND))
+                                {
+                                    unsigned int destNum = next->bbJumpDest->bbNum;
+                                    if ((destNum >= top->bbNum) && (destNum <= bottom->bbNum) &&
+                                        !loopBlocks.isMember(destNum))
+                                    {
+                                        // Reversing this branch out of block `next` could confuse this algorithm, so
```

Comment:
Those branches would still be reversed if we moved the blocks anywhere after `next`.

The clause where we don't search into a different EH region is one where we could later find another viable candidate (if the blocks after the loop fall through into a new `try` and then back out of it), but I'm not sure it's worth the complexity to keep searching for those cases.

Regarding "if you find one, use it, otherwise just insert after BOTTOM", were you just sketching out how an algorithm that kept walking would work, or do you have a reason to think that BOTTOM is a better insertion point than the last legal block?